src/realm/object-store/impl/epoll/external_commit_helper.cpp

```diff
@@ -49,31 +49,53 @@ using namespace realm::_impl;
     } while (0)
 
 namespace {
-// Write a byte to a pipe to notify anyone waiting for data on the pipe
-void notify_fd(int fd)
+
+// Make writing to the pipe return -1 when there is no data to read, or no space in the buffer to write to, rather
+// than blocking.
+void make_non_blocking(int fd)
+{
+    int ret = fcntl(fd, F_SETFL, O_NONBLOCK);
+    if (ret == -1) {
+        throw std::system_error(errno, std::system_category());
+    }
+}
+
+// Write a byte to a pipe to notify anyone waiting for data on the pipe.
+// But first consume all bytes in the pipe, since linux may only notify on transition from not ready to ready.
+// If a process dies after reading but before writing, it can consume a pending notification, and possibly prevent
+// other processes from observing it. This is a transient issue and the next notification will work correctly.
+void notify_fd(int fd, bool read_first = true)
 {
     while (true) {
+        if (read_first) {
+            while (true) {
+                uint8_t buff[1024];
+                ssize_t actual = read(fd, buff, sizeof(buff));
+                if (actual == 0) {
+                    break; // Not sure why we would see EOF here, but defer error handling to the writer.
```

Comment:
Isn't reading zero bytes the expected result of the first call to this function?

Reply:
No, that would return EAGAIN or EWOULDBLOCK. Zero is only returned for end-of-file.